### DISCUSSION TITLE
LDAP group search issue

### DISCUSSION THREAD

**Discussion by ab1g17** [inventree/InvenTree#8365] [Q&A] [answered]
Hello everyone,

My current LDAP config (obfuscated):
INVENTREE_LDAP_ENABLED=True
INVENTREE_LDAP_SERVER_URI=ldap://myldapserver
INVENTREE_LDAP_SEARCH_FILTER_STR=(sAMAccountName=%(user)s)
INVENTREE_LDAP_BIND_DN='CN=serviceaccount,OU=Users,DC=eu,DC=corp,DC=com'
INVENTREE_LDAP_BIND_PASSWORD='pass'
INVENTREE_LDAP_START_TLS=False
INVENTREE_LDAP_DEBUG=True
INVENTREE_LDAP_USER_DN_TEMPLATE=
INVENTREE_LDAP_BASE_DN='DC=corp,DC=com'
INVENTREE_LDAP_USER_ATTR_MAP={"first_name": "givenName", "last_name": "Surname", "email": "EmailAddress"}
INVENTREE_LDAP_GROUP_SEARCH=OU=Users,OU=Site,dc=eu,dc=corp,dc=com
INVENTREE_LDAP_REQUIRE_GROUP=Site Catalog Tool Users

For any username I try to login as (even ones that dont exist in AD at all) the error I see in the logs is:

Binding as CN=serviceaccount,OU=Users,DC=eu,DC=domain,DC=com
Binding as CN=serviceaccount,OU=Users,DC=eu,DC=domain,DC=com
 is not a member of site catalog tool users
 is not a member of site catalog tool users
Authentication failed for john.smith: user does not satisfy AUTH_LDAP_REQUIRE_GROUP
Authentication failed for john.smith: user does not satisfy AUTH_LDAP_REQUIRE_GROUP

Is there something wrong with the group search config?

Things I have tried:
checked spelling of group names,
making sure user is indeed part of the group,
using DN of group instead of just a string of the name
made sure the serviceaccount has rights to perform search of groups

**Comment by ab1g17** [comment]
@wolflu05 maybe you could shed some light here?

**Reply by wolflu05** [reply]
Hi, sorry that I can't really help here, as I'm not using ldap in a setup anymore. But I can give you some links to documentations:

Basically 
- LDAP group support was added by https://github.com/inventree/InvenTree/pull/6003
- basically that's just forwarding env/config variables to an external module named Django-auth-ldap, see https://github.com/inventree/InvenTree/blob/6953640161a24376afb02e77c26add295b7a7b0d/src/backend/InvenTree/InvenTree/settings.py#L361-L458
- Documentation for that module is available at https://django-auth-ldap.readthedocs.io/en/latest/groups.html

Maybe that's for some help of you.

**Reply by ab1g17** [reply]
apparently the code is looking for "uniqueMember" attribute when searching the groups, and my AD environment uses the "member" attribute to store the users as members of groups. Any idea if there is a way to fix this without changing the code?

**Reply by wolflu05** [reply, marked as answer]
Are you referring to 
https://github.com/inventree/InvenTree/blob/6953640161a24376afb02e77c26add295b7a7b0d/src/backend/InvenTree/InvenTree/settings.py#L445 

And you want to have `'(objectClass=groupOfNames)'` ? Then the best way would be to add a setting for this as well. The best would be if you could submit a PR for this then.

**Reply by wolflu05** [reply]
#8475 may be related??

**Reply by jacobfelknor** [reply]
Yes, if  #8475 is merged, this should address this issue. For example, in my environment, I specified the following new options

```yaml
INVENTREE_LDAP_GROUP_OBJECT_CLASS: group
INVENTREE_LDAP_GROUP_TYPE_CLASS: NestedMemberDNGroupType
INVENTREE_LDAP_GROUP_TYPE_CLASS_ARGS: member
INVENTREE_LDAP_GROUP_TYPE_CLASS_KWARGS: '{"name_attr": "cn"}'
````

**Reply by ab1g17** [reply]
@jacobfelknor could you please share a working full config?

I have  the following options in my .env:

INVENTREE_LDAP_ENABLED=True
INVENTREE_LDAP_SERVER_URI=ldap://my-domain-controller
INVENTREE_LDAP_SEARCH_FILTER_STR=(sAMAccountName=%(user)s)
INVENTREE_LDAP_START_TLS=False
INVENTREE_LDAP_DEBUG=True
INVENTREE_LDAP_USER_DN_TEMPLATE=
INVENTREE_LDAP_BASE_DN='DC=domain,DC=com'

INVENTREE_LDAP_GROUP_SEARCH='DC=subdomain,DC=domain,DC=com'
INVENTREE_LDAP_GROUP_OBJECT_CLASS=group
INVENTREE_LDAP_GROUP_TYPE_CLASS=NestedMemberDNGroupType
INVENTREE_LDAP_GROUP_TYPE_CLASS_ARGS=member
INVENTREE_LDAP_REQUIRE_GROUP='Name of my group'



My logs say:
2025-01-16 11:14:10,382 DEBUG Binding as None
Invoking search_s('DC=eu,DC=domain,DC=com', 2, '(&(objectClass=group)(|(member=)))')
2025-01-16 11:14:10,383 DEBUG Invoking search_s('DC=eu,DC=domain,DC=com', 2, '(&(objectClass=group)(|(member=)))')
Caught LDAPError while authenticating: OPERATIONS_ERROR({'msgtype': 101, 'msgid': 3, 'result': 1, 'desc': 'Operations error', 'ctrls': [], 'info': '000004DC: LdapErr: DSID-0C090C90, comment: In order to perform this operation a successful bind must be completed on the connection., data 0, v4563'})
2025-01-16 11:14:10,383 WARNING Caught LDAPError while authenticating: OPERATIONS_ERROR({'msgtype': 101, 'msgid': 3, 'result': 1, 'desc': 'Operations error', 'ctrls': [], 'info': '000004DC: LdapErr: DSID-0C090C90, comment: In order to perform this operation a successful bind must be completed on the connection., data 0, v4563'})
10.137.186.4 - - [16/Jan/2025:11:14:10 +0000] "POST /accounts/login/ HTTP/1.1" 200 8681 "http://URL/accounts/login/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36"

**Reply by jacobfelknor** [reply]
My full LDAP sanitized configuration:

```yml
# LDAP Support
INVENTREE_LDAP_ENABLED: True
INVENTREE_LDAP_DEBUG: False
INVENTREE_LDAP_SERVER_URI: "ldap://<redacted>:389"
INVENTREE_LDAP_START_TLS: False
INVENTREE_LDAP_AUTH_SEARCH_BASE: "<redacted>"
INVENTREE_LDAP_BIND_DN: "<redacted>"
INVENTREE_LDAP_BIND_PASSWORD: "<redacted>"
INVENTREE_LDAP_SEARCH_BASE_DN: "<redacted>"
INVENTREE_LDAP_SEARCH_FILTER_STR: "(&(objectCategory=user)(sAMAccountName=%(user)s))"
INVENTREE_LDAP_MIRROR_GROUPS: True
INVENTREE_LDAP_GROUP_SEARCH: "<redacted>"
# NEW OPTIONS
INVENTREE_LDAP_GROUP_OBJECT_CLASS: group
INVENTREE_LDAP_GROUP_TYPE_CLASS: NestedMemberDNGroupType
INVENTREE_LDAP_GROUP_TYPE_CLASS_ARGS: member
INVENTREE_LDAP_GROUP_TYPE_CLASS_KWARGS: '{"name_attr": "cn"}'
```

Note that the specifics will be highly tailored to your environment, so I'm not sure how much help this is

**Reply by ab1g17** [reply]
Won't work in my envirnoment

If I use INVENTREE_LDAP_BIND_DN to search, it forces me to use INVENTREE_LDAP_SEARCH_BASE_DN to the very specific DN of users. But not all my users share the same exact DN structure. Some are in one subdomain/OU, some are in others

If I dont use INVENTREE_LDAP_BIND_DN and just try to search as the user being authenticated as I was already, then group search doesnt work

This app's ldap implementation seems to be very simplistic and wont work in an environment of high complexity with lots of OUs and subdomains

**Reply by jacobfelknor** [reply]
We pass through the settings to [`django-auth-ldap`](https://django-auth-ldap.readthedocs.io/en/latest/).

Feel free to suggest any changes to make it work for your needs

**Reply by ab1g17** [reply]
Basically I just need the code to not ask for a specific DN template when I use the INVENTREE_LDAP_BIND_DN option and bind to a service account to search.
Not all environments will have a single OU for all users. The app should know to look for the users in the base DN (dc=domain,dc=com), whatever final DN the user might be in
This is how I managed to get plain authenticating to work in my environment, but I could only get it to work without binding to any account to search. Basically I use:
INVENTREE_LDAP_USER_DN_TEMPLATE=
INVENTREE_LDAP_BASE_DN='DC=domain,DC=com'

notice that I had to leave INVENTREE_LDAP_USER_DN_TEMPLATE empty, and this is what forces the app to allow any final DN for the user. This is not explained in the documentation, I just managed to find this workaround after many attempts at different things (including not providing that option at all)

However, if I also want the new group search feature enabled, this workaround is no longer valid, because I have to bind to an account to search. So I have to use INVENTREE_LDAP_SEARCH_BASE_DN instead of INVENTREE_LDAP_BASE_DN, and I have to give the specific DN, I cant just leave "dc=domain,dc=com" like before. The leaving it empty trick does not work in this case.

If I am making a mistake let me know, but I've tried all possible combinations without success

**Reply by matmair** [reply]
this sounds reasonable; I can not promise anything but we can look into it.
@ab1g17 can you please open a feature request describing how the user experience should change?

**Reply by ab1g17** [reply]
done:
https://github.com/inventree/InvenTree/discussions/8974
